### PR TITLE
Overhaul put_resource logic.

### PR DESCRIPTION
**Pull request by rohinb2**
Introduce a generic format and precedence for calls into `EnvServlet`.

All functions will take in `data` and `serialization`. They can put whatever
they want in the `data` field and unpack them as tuples before and after, as
done in this PR. This will support pickle and json serialization.

The other parameters to `EnvServlet` functions will be explicit.

The `EnvServlet` should also only call `local` ObjStore functions, and the
ObjStore should deal with all the outside-of-env logic and call the right Env.

### CODE REVIEW

**Comment by rohinb2** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #356** <a href="https://app.graphite.dev/github/pr/run-house/runhouse/356?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/run-house/runhouse/356?utm_source=stack-comment).

File: runhouse/servers/env_servlet.py [hunk 1]
```diff
@@ -19,18 +20,53 @@
...
+                )
+            else:
+                return Response(
+                    output_type=OutputType.SUCCESS,
+                )
+        except Exception as e:
+            logger.exception(e)
+            return Response(
+                error=pickle_b64(e),
+                traceback=pickle_b64(traceback.format_exc()),
```
Comment (dongreenberg) [inline runhouse/servers/env_servlet.py:63]: Shouldn't this also use the serialization option passed above?

File: runhouse/servers/http/http_server.py [hunk 2]
```diff
@@ -308,31 +308,23 @@ def update_settings(request: Request, message: ServerSettings) -> Response:
...
-                )
-
-        return HTTPServer.call_in_env_servlet(
-            "put_resource",
-            [message],
-            env=message.env,
-            create=True,
-            lookup_env_for_name=message.key,
-        )
+    def put_resource(request: Request, params: PutResourceParams):
```
Comment (dongreenberg) [inline runhouse/servers/http/http_server.py:311]: Oo pydantic validation is this the ritz?

File: runhouse/servers/http/http_server.py [hunk 3]
```diff
@@ -308,31 +308,23 @@ def update_settings(request: Request, message: ServerSettings) -> Response:
...
+                serialized_data=params.serialized_data,
+                serialization=params.serialization,
+                env_name=env_name,
+            )
+        except Exception as e:
+            logger.exception(e)
+            HTTPServer.register_activity()
+            return Response(
+                error=pickle_b64(e),
+                traceback=pickle_b64(traceback.format_exc()),
```
Comment (dongreenberg) [inline runhouse/servers/http/http_server.py:325]: I continue to wonder out loud if we need to serialize these. Most likely not. Strings are probably fine.

Comment (dongreenberg) [reply runhouse/servers/http/http_server.py:325]: In the HTTPServer specifically. I don't know for the EnvServlet and obj store.

File: runhouse/servers/http/http_utils.py [hunk 4]
```diff
@@ -26,6 +27,12 @@ class ServerSettings(BaseModel):
     flush_auth_cache: Optional[bool] = None
 
 
+class PutResourceParams(BaseModel):
+    serialized_data: Any
+    serialization: Optional[str] = None
+    env_name: Optional[str] = None
```
Comment (dongreenberg) [inline runhouse/servers/http/http_utils.py:33]: Just noting this is quite generic. I'd expect that if we had pydantic types for each route we'd go more specific (e.g. not bundle into serialized_data), or we reuse a type like this for several things which share these inputs.

File: runhouse/servers/obj_store.py [hunk 5]
```diff
@@ -603,3 +603,95 @@ def get_obj_refs_dict(self, d: Dict[Any, Any]):
...
+        serialization: Optional[str] = None,
+        env_name: Optional[str] = None,
+    ) -> "Response":
+        from runhouse.servers.http.http_utils import deserialize_data
+
+        if env_name is None:
+            if self.servlet_name is None:
+                raise ObjStoreError("No env name provided and no servlet name set.")
+            else:
+                env_name = self.servlet_name
```
Comment (dongreenberg) [inline runhouse/servers/obj_store.py:622]: If we're inside this env, should we just call self.put_resource_local directly? No reason to serialize, deserialize, and add a ray call to the servlet if we're already in the right env process?

File: tests/test_servers/test_http_server.py [hunk 6]
```diff
@@ -58,7 +58,9 @@ def test_put_resource(self, http_client, blob_data, cluster):
         resource = rh.blob(data=blob_data, system=cluster)
         data = pickle_b64((resource.config_for_rns, state, resource.dryrun))
         response = http_client.post(
-            "/resource", json={"data": data}, headers=rns_client.request_headers()
+            "/resource",
+            json=PutResourceParams(serialized_data=data, serialization="pickle").dict(),
```
Comment (dongreenberg) [inline tests/test_servers/test_http_server.py:62]: 👁️👄👁️

**Review by dongreenberg — APPROVED** [review]
Overall excellent, much cleaner!

File: runhouse/servers/env_servlet.py (hunk 1 above)
Comment (rohinb2) [reply runhouse/servers/env_servlet.py:63]: currently we deserialize all exceptions the same as using pickle, client side

File: runhouse/servers/http/http_server.py (hunk 3 above)
Comment (rohinb2) [reply runhouse/servers/http/http_server.py:325]: I can remove this later

File: runhouse/servers/http/http_utils.py (hunk 4 above)
Comment (rohinb2) [reply runhouse/servers/http/http_utils.py:33]: It'd be harder to make the serialization arg general then

File: runhouse/servers/obj_store.py (hunk 5 above)
Comment (rohinb2) [reply runhouse/servers/obj_store.py]: fixed

**Comment by rohinb2** [comment]
### Merge activity

* **Jan 19, 11:52 AM**: @rohinb2 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/run-house/runhouse/356).

**Pull request merged**